src/plonk/lookup/prover.rs

```diff
@@ -176,6 +176,200 @@ impl Argument {
+                *lookup_product *= &(gamma + permuted_table_value);
+            }
+        });
+
+        // Batch invert to obtain the denominators for the lookup product
+        // polynomials
+        lookup_product.iter_mut().batch_invert();
+
+        // Finish the computation of the entire fraction by computing the numerators
+        // (a_1(X) + \theta a_2(X) + ... + \beta) (s_1(X) + \theta s_2(X) + ... + \gamma)
+        // Compress unpermuted input columns
+        let mut input_term = vec![C::Scalar::zero(); params.n as usize];
+        for unpermuted_input_value in unpermuted_input_values.iter() {
+            parallelize(&mut input_term, |input_term, start| {
+                for (input_term, input_value) in input_term
+                    .iter_mut()
+                    .zip(unpermuted_input_value[start..].iter())
+                {
+                    *input_term *= &theta;
+                    *input_term += input_value;
+                }
+            });
+        }
+
+        // Compress unpermuted table columns
+        let mut table_term = vec![C::Scalar::zero(); params.n as usize];
+        for unpermuted_table_value in unpermuted_table_values.iter() {
+            parallelize(&mut table_term, |table_term, start| {
+                for (table_term, fixed_value) in table_term
+                    .iter_mut()
+                    .zip(unpermuted_table_value[start..].iter())
+                {
+                    *table_term *= &theta;
+                    *table_term += fixed_value;
+                }
+            });
+        }
+
+        // Add \beta and \gamma offsets
+        parallelize(&mut lookup_product, |product, start| {
+            for ((product, input_term), table_term) in product
+                .iter_mut()
+                .zip(input_term[start..].iter())
+                .zip(table_term[start..].iter())
+            {
+                *product *= &(*input_term + &beta);
+                *product *= &(*table_term + &gamma);
+            }
+        });
+
+        // The product vector is a vector of products of fractions of the form
+        //
+        // (a_1(\omega^i) + \theta a_2(\omega^i) + ... + \beta)(s_1(\omega^i) + \theta s_2(\omega^i) + ... + \gamma)/
+        // (a'(\omega^i) + \beta) (s'(\omega^i) + \gamma)
+        //
+        // where a_j(\omega^i) is the jth input column in this lookup,
+        // a'j(\omega^i) is the permuted input column,
+        // s_j(\omega^i) is the jth table column in this lookup,
+        // s'(\omega^i) is the permuted table column,
+        // and i is the ith row of the column.
+
+        // Compute the evaluations of the lookup product polynomial
+        // over our domain, starting with z[0] = 1
+        let mut z = vec![C::Scalar::one()];
+        for row in 1..(params.n as usize) {
+            let mut tmp = z[row - 1];
+            tmp *= &lookup_product[row];
+            z.push(tmp);
+        }
```

Comment:
Yeah, `let mut z = vec![C::Scalar::one()];` initializes the vector with the first element being one, with the intention to show that the product of the numerators is the inverse of the product of the denominators. Side note, this code should probably be altered so that the z vector is initialized with the correct capacity (`params.n`) to reduce allocations.

Reply:
Mmk, so `z` ends up containing in its last entry the product of all parts of `lookup_product` except the first? It's not clear to me where that last part is accounted for.

Re: reducing allocations, AFAIK the chain-and-collects I propose above and in other comments should also achieve that. I should dig into `collect` to check this...